File: android/core/data/src/main/kotlin/app/hapi/data/store/ChatHistoryPagingState.kt
package app.hapi.data.store

import app.hapi.protocol.window.OlderLoadOutcome

/** A page remains in flight until its version has been laid out. UI-thread owned. */
data class ChatHistoryPagingState(
    val phase: Phase = Phase.Idle,
    val generation: Long = 0,
    val failures: Int = 0,
    val pagesWithoutProgress: Int = 0,
) {
    sealed interface Phase {
        data object Idle : Phase
        data object Loading : Phase
        data class AwaitingLayout(val historyVersion: Long) : Phase
        data object Retrying : Phase
        data object Failed : Phase
        data object Paused : Phase
        data object Exhausted : Phase
    }

    fun begin(): ChatHistoryPagingState =
        if (phase == Phase.Idle) copy(phase = Phase.Loading, generation = generation + 1) else this

    fun received(result: OlderLoadOutcome, request: Long): ChatHistoryPagingState {
        if (generation != request || phase != Phase.Loading) return this
        return when (result) {
            is OlderLoadOutcome.Applied -> copy(phase = Phase.AwaitingLayout(result.historyVersion), failures = 0)
            is OlderLoadOutcome.Failed -> copy(
                phase = if (failures < 2) Phase.Retrying else Phase.Failed,
                failures = failures + 1,
            )
            is OlderLoadOutcome.Stopped -> copy(phase = when (result.reason) {
                OlderLoadOutcome.StopReason.Exhausted -> Phase.Exhausted
                OlderLoadOutcome.StopReason.CursorDidNotAdvance -> Phase.Paused
                else -> Phase.Idle
            })
        }
    }

    val retryDelayMillis: Long? get() = if (phase != Phase.Retrying) null else if (failures == 1) 500 else 1500

    fun retryElapsed(request: Long): ChatHistoryPagingState =
        if (generation == request && phase == Phase.Retrying) copy(phase = Phase.Idle) else this

    fun laidOut(historyVersion: Long, madeProgress: Boolean): ChatHistoryPagingState {
        val waiting = phase as? Phase.AwaitingLayout ?: return this
        if (historyVersion < waiting.historyVersion) return this
        val emptyPages = if (madeProgress) 0 else pagesWithoutProgress + 1
        return copy(phase = if (emptyPages >= 3) Phase.Paused else Phase.Idle, pagesWithoutProgress = emptyPages)
    }

    fun resume(): ChatHistoryPagingState =
        if (phase == Phase.Failed || phase == Phase.Paused) copy(phase = Phase.Idle, failures = 0, pagesWithoutProgress = 0) else this

    fun cancel(): ChatHistoryPagingState = ChatHistoryPagingState(generation = generation + 1)

    fun refreshAvailability(hasMore: Boolean): ChatHistoryPagingState =
        if (hasMore && phase == Phase.Exhausted) copy(phase = Phase.Idle) else this
}

File: ios/Packages/HapiKit/Sources/HapiClient/Chat/ChatHistoryPagingState.swift
import HapiProtocol

/// One pagination run includes its first rendered layout, not just the HTTP
/// request. Geometry may submit demand repeatedly; only an idle run starts.
/// Owned by the chat's main-thread coordinator. No timers or UI dependencies.
public struct ChatHistoryPagingState: Equatable, Sendable {
    public enum Phase: Equatable, Sendable {
        case idle, loading, awaitingLayout(Int), retrying, failed, paused, exhausted
    }

    public private(set) var phase: Phase = .idle
    public private(set) var generation = 0
    public private(set) var failures = 0
    public private(set) var pagesWithoutProgress = 0

    public init() {}

    public mutating func begin() -> Int? {
        guard phase == .idle else { return nil }
        generation += 1
        phase = .loading
        return generation
    }

    /// Returns the bounded retry delay, in milliseconds, if one is needed.
    @discardableResult
    public mutating func received(_ result: OlderLoadOutcome, generation: Int) -> Int? {
        guard self.generation == generation, phase == .loading else { return nil }
        switch result {
        case .applied(let version, _, _):
            phase = .awaitingLayout(version)
            failures = 0
        case .failed:
            failures += 1
            phase = failures <= 2 ? .retrying : .failed
            return failures == 1 ? 500 : failures == 2 ? 1500 : nil
        case .stopped(let reason):
            switch reason {
            case .exhausted: phase = .exhausted
            case .cursorDidNotAdvance: phase = .paused
            default: phase = .idle
            }
        }
        return nil
    }

    @discardableResult
    public mutating func retryElapsed(generation: Int) -> Bool {
        guard self.generation == generation, phase == .retrying else { return false }
        phase = .idle
        return true
    }

    /// Also called for pages with no new cells: their version still needs an
    /// acknowledgement, otherwise hidden-only pages deadlock the loader.
    @discardableResult
    public mutating func laidOut(historyVersion: Int, madeProgress: Bool) -> Bool {
        guard case .awaitingLayout(let target) = phase, historyVersion >= target else { return false }
        pagesWithoutProgress = madeProgress ? 0 : pagesWithoutProgress + 1
        phase = pagesWithoutProgress >= 3 ? .paused : .idle
        return true
    }

    public mutating func resume() {
        guard phase == .failed || phase == .paused else { return }
        failures = 0
        pagesWithoutProgress = 0
        phase = .idle
    }

    /// Tail-window trimming can make older pages available again. Errors and
    /// no-progress pauses still require an explicit reader action.
    public mutating func refreshAvailability(hasMore: Bool) {
        if hasMore, phase == .exhausted { phase = .idle }
    }

    public mutating func cancel() {
        generation += 1
        phase = .idle
        failures = 0
        pagesWithoutProgress = 0
    }
}
